cores/esp8266/uart.c

```diff
@@ -91,6 +91,33 @@ size_t uart_resize_rx_buffer(uart_t* uart, size_t new_size)
     return uart->rx_buffer->size;
 }
 
+inline size_t uart_rx_buffer_available(uart_t* uart) {
+    if(uart->rx_buffer->wpos < uart->rx_buffer->rpos) {
+      return (uart->rx_buffer->wpos + uart->rx_buffer->size) - uart->rx_buffer->rpos;
+    }
+    return uart->rx_buffer->wpos - uart->rx_buffer->rpos;
+}
+
+inline size_t uart_rx_fifo_available(uart_t* uart) {
+    return (USS(uart->uart_nr) >> USRXC) & 0x7F;
+}
+
+// Copy all the rx fifo bytes that fit into the rx buffer
+inline void uart_rx_copy_fifo_to_buffer(uart_t* uart) {
+    while(uart_rx_fifo_available(uart)){
+        size_t nextPos = (uart->rx_buffer->wpos + 1) % uart->rx_buffer->size;
+        if(nextPos != uart->rx_buffer->rpos) {
```

Comment:
Please don't use inverted logic. Instead, use direct logic to detect an early return or break, and leave the 3 lines of code free-standing after that.
E.g.: 
if(nextPos == rpos)
    return;
uint8_t data = ...

Reply:
So the code would look like this:
'
inline void uart_rx_copy_fifo_to_buffer(uart_t* uart) {
    while(uart_rx_fifo_available(uart)){
        size_t nextPos = (uart->rx_buffer->wpos + 1) % uart->rx_buffer->size;
        if(nextPos == uart->rx_buffer->rpos) {
            // Stop copying if rx buffer is full
            break;
        }
        uint8_t data = USF(uart->uart_nr);
        uart->rx_buffer->buffer[uart->rx_buffer->wpos] = data;
        uart->rx_buffer->wpos = nextPos;
    }
}
'

I agree that code looks better, but I do have a concern.  The line that initizes data with a value from the fifo also drains the fifo by 1 byte.  So that line of code really can't be executed in the case we want to execute the break.  I'm not sure if the compiler is allowed to rearrange that line to be before the if.  If it can do that with any optimization level then we'd have a bug.  So while it isn't quite as clean I'd suggest putting that code in an else (I can keep the current order though so the break is first).  However, if you are sure this is legal then I can make that change.

Reply:
Nevermind.  It's late and I was overthinking this.  Of course a compile can't reorder this if it might affect results.  I'll test this change with my test case and if it passes will push the change.